Sources/AsyncAlgorithms/MultiProducerSingleConsumerChannel/MultiProducerSingleConsumerChannel.swift

```diff
@@ -0,0 +1,489 @@
+            ///
+            /// - Parameters:
+            ///   - low: When the number of buffered elements drops below the low watermark, producers will be resumed.
+            ///   - high: When the number of buffered elements rises above the high watermark, producers will be suspended.
+            public static func watermark(low: Int, high: Int) -> BackpressureStrategy {
+                .init(
+                    internalBackpressureStrategy: .watermark(
+                        .init(low: low, high: high, waterLevelForElement: nil)
+                    )
+                )
+            }
+
+            /// A backpressure strategy using a high and low watermark to suspend and resume production respectively.
+            ///
+            /// - Parameters:
+            ///   - low: When the number of buffered elements drops below the low watermark, producers will be resumed.
+            ///   - high: When the number of buffered elements rises above the high watermark, producers will be suspended.
+            ///   - waterLevelForElement: A closure used to compute the contribution of each buffered element to the current water level.
+            ///
+            /// - Note, `waterLevelForElement` will be called on each element when it is written into the source and when
+            /// it is consumed from the channel, so it is recommended to provide an function that runs in constant time.
+            public static func watermark(
+                low: Int,
+                high: Int,
+                waterLevelForElement: @escaping @Sendable (Element) -> Int // TODO: In the future this should become sending
+            ) -> BackpressureStrategy {
+                .init(
+                    internalBackpressureStrategy: .watermark(
+                        .init(low: low, high: high, waterLevelForElement: waterLevelForElement)
+                    )
+                )
+            }
+
+            /// An unbounded backpressure strategy.
+            ///
+            /// - Important: Only use this strategy if the production of elements is limited through some other mean. Otherwise
+            /// an unbounded backpressure strategy can result in infinite memory usage and open your application to denial of service
+            /// attacks.
+            public static func unbounded() -> BackpressureStrategy {
+                .init(
+                    internalBackpressureStrategy: .unbounded(.init())
+                )
+            }
+        }
+
+        /// A type that indicates the result of sending elements to the source.
+        public enum SendResult: ~Copyable, Sendable {
+            /// A token that is returned when the channel's backpressure strategy indicated that production should
+            /// be suspended. Use this token to enqueue a callback by  calling the ``enqueueCallback(_:)`` method.
+            public struct CallbackToken: Sendable {
```

Comment:
often tokens need to be hashable or at the bare min equatable; that might be worthwhile t consider.